node/logging-and-metrics-for-user-code.md

```diff
@@ -0,0 +1,184 @@
+## Logging and Metrics in user Workflow code
+
+A hard requirement for our logging and metrics solution for user code is that we do not enforce a specific tool or convention and instead let our users use their existing tools.<br/>
+For Activities, there is not an issue because users can run anything in an Activity. In Workflows OTOH, only deterministic code can run and metrics and logs should usually be dropped when a Workflow is replaying. Additionally the fact that the Node SDK runs Workflows in v8 isolates, mean that there's no way to do IO or anything that interacts with the world without explicit support from the SDK.
+
+### IO in Workflow code
+
+There are a few options for doing IO inside a Workflow's isolate:
+
+1. Load a binary (c++) module into the isolate, the module is not isolated the way JS code is
+2. Inject a function reference from the main NodeJS isolate into a Workflow isolate
+3. Accumulate data in the isolate (e.g when collecting commands to complete an activation) and pass it into a function in the main NodeJS isolate
```

Comment:
Is there any concern about reliability with this option? Such as a situation where the logs/metrics get collected but don't have an opportunity to flush. This could make it very challenging to debug unexpected errors that are killing a thread.

Reply:
Good point, this is something that might happen in case the isolate runs out of memory but in that case logs will probably not work anyways, I think the isolate will die.
I thought there might be an issue of WF code getting stuck e.g goes into an infinite loop but we can mitigate this problem by enforcing a time limit of execution in the isolate, in case execution times out we will be able to extract the logs.

Reply:
I've stated these concerns in the document, resolving.